### ISSUE TITLE
Token refresh body is empty

### ISSUE THREAD

**Issue by caiusCitiriga** [transistorsoft/cordova-background-geolocation-lt#1071]
## Your Environment
* Plugin version: 3.4.2
* Platform: Android
* OS version: 9
* Device manufacturer / model: OnePlus 5T
* Cordova version (`cordova -v`): 9.0.0
* Cordova platform version (`cordova platform ls`): android 8.1.0
* Plugin config provided to `#ready`:
```typescript
const jwt = await this.storage.get(StorageKeys.authToken);
const refToken = await this.storage.get(StorageKeys.refreshToken);
BackgroundGeolocation.ready({
    debug: true,
    startOnBoot: true,
    distanceFilter: 10,
    reset: resetConfig,
    stopOnTerminate: false,
    foregroundService: true,
    geofenceModeHighAccuracy: true,
    logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
    desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
    authorization: {
        strategy: 'JWT',
        accessToken: jwt,
        refreshToken: refToken,
        refreshPayload: { refreshToken: '{refreshToken}' },
        refreshUrl: `${environment.ApiEndpoint}/auth/refresh`,
    },
    url: `${environment.ApiEndpoint}tracking/position/logs`,
}, async (state) => {
    this.startBackgroundGeolocation(mode);
    resolver(state);
    this.geolocationReady$.next(true);
});
```

## Expected Behavior
The plugin should send the correct body to the server when trying to refresh the token. 

## Actual Behavior
The server receives an empty body on the refresh endpoint: 
```text
{
  "data": {
    "error": "jwt must be provided",
    "request": {
      "headers": {
        "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1aWQiOiJ2dHBZM0ZQVGZTYnprZzlzcG93UEh2akkyV2YxIiwiaWF0IjoxNTc4OTM2MTkyLCJleHAiOjE1ODE1MjgxOTIsImlzcyI6InNzLWlkZW50aXR5In0.OxbAVG2A8hyUgLlaZjSLoBKpm8vthStbc7RaQ7VuZyY",
        "content-type": "multipart/form-data; boundary=8557930b-f333-48cf-b1b3-be6bc6cfcbdd",
        "content-length": "364",
        "host": "api.dev.testapp.co",
        "accept-encoding": "gzip",
        "user-agent": "okhttp/3.12.8",
        "x-forwarded-port": "443",
        "x-forwarded-proto": "https",
        "x-forwarded-for": "79.6.56.105",
        "connection": "close",
        "accept-language": "en"
      },
      "url": "auth/refresh",
      "body": {}
    }
  },
  "trace": "authController.refreshToken",
  "context": "LoggingInterceptor",
  "message": "<--OUT ERR [my751wxbpu] [POST auth/refresh]",
  "ts": 1578991767839,
  "requestUid": "my751wxbpu"
}
```
## Debug logs
A strange message can be found in the logs of the libarary:

*`Value jwt of type java.lang.String cannot be converted to JSONObject`*

```<!-- Syntax highlighting:  DO NOT REMOVE -->
╔═════════════════════════════════════════════
║ HTTP Service (count: 3)
╠═════════════════════════════════════════════

01-13 19:25:41.651 INFO [BackgroundTaskManager onStartJob] ⏳ startBackgroundTask: 41
01-13 19:25:41.654 DEBUG [b first] 
  ✅  Locked 1 records
01-13 19:25:41.655 INFO [HttpService a] 
  🔵  HTTP POST: 70c5bcde-7e1a-42c5-a5bd-e74fade7b560
01-13 19:25:41.753 DEBUG [ActivityRecognitionService onDestroy] 
  🔴  ActivityRecognitionService destroyed
01-13 19:25:43.381 WARN [HttpService$c onResponse] 
  ⚠️  Response: 401, Unauthorized
01-13 19:25:43.386 DEBUG [b unlock] 
  ✅  UNLOCKED: 70c5bcde-7e1a-42c5-a5bd-e74fade7b560
01-13 19:25:43.387 INFO [BackgroundTaskManager$Task stop] ⏳ stopBackgroundTask: 41
01-13 19:25:43.399 INFO [BackgroundTaskManager onStartJob] ⏳ startBackgroundTask: 42
01-13 19:25:43.400 DEBUG [BackgroundTaskService onDestroy] 
01-13 19:25:44.019 WARN [TSAuthorization _onFailure] 
  ⚠️  🔑 Value jwt of type java.lang.String cannot be converted to JSONObject
01-13 19:25:44.020 INFO [BackgroundTaskManager$Task stop] ⏳ stopBackgroundTask: 42
01-13 19:30:01.086 INFO [TSScheduleManager oneShot] 
  ⏰ Scheduled OneShot: TERMINATE_EVENT in 10000ms (jobID: -1708771588)
01-13 19:30:11.101 INFO [ScheduleEvent onOneShot] 
```

</details>

I've double checked everything, but I can't figure out what's wrong... Any clue Chris?

**Comment by christocracy** [comment]
In the react-native version, another customer pointed out to me that i was using the wrong `Content-Type` in the refreshUrl request (`multipart/form`).
```
"content-type": "multipart/form-data; boundary=8557930b-f333-48cf-b1b3-be6bc6cfcbdd",
```

This has been changed to `application/x-www-form-urlencoded`.  It will be available in the `#dev` branch today.

**Comment by caiusCitiriga** [comment]
Hey @christocracy! Yeah, I've noticed that, in fact, the server expects `application/json` as the rest of the calls. Wouldn't it be possible to set this manually? 

For now the workaround for the refresh is to intercept the 401 status on each httpEvent, and then manually refresh the token. But I'm wondering if this callback executes even if the app is in background, or worse, killed...?

**Comment by christocracy** [comment]
The OAuth 2.0 RFC requires the refresh-url accept `application/x-www-form-urlencoded`.
https://tools.ietf.org/html/rfc6749#section-6

It's not easy to allow multiple content-type.  It's not just the header which needs to be changed, it's the Obj-c / Java HTTP request code which would need to operate for multiple request types.

**Comment by caiusCitiriga** [comment]
Ok, I see. I'll modify the server then. Once I'll be sure that it will work, I'll let you know and close the issue. 

For now, can you tell me if the onHttpEvent callback gets executed even if the app is backgrounded/killed?

In the meanwhile thank you.

**Comment by christocracy** [comment]
> onHttpEvent callback gets executed even if the app is backgrounded/killed?

Not in terminated state.  See Wiki “Android Headless Mode” for handling events in terminated state.

**Comment by caiusCitiriga** [comment]
Hey @christocracy I'm trying to checkout the #dev branch to see if the `application/x-www-form-urlencoded` header works on my server, but I can't find the branch you're talking about on this repo. Am I missing something?

**Comment by christocracy** [comment]
Latest updates are available only in the private repo.

The dev branch in private has been merged to master as 3.5.0.

This public repo is mirrored from private repo only about every 3 months.

**Comment by caiusCitiriga** [comment]
Hey @christocracy thank you for the clarification. I'm closing this since I don't think it's the case to keep it open since we cannot implement the change immediately. But I'm pretty confident that when the server will support the proper encoding everything will start running as expected. In the meantime we've found another workaround. 

If you agree, I'll reopen this same issue whenever and if any problem comes out.
